packages/devtools_app/lib/src/debugger/flutter/debugger_controller.dart

```diff
@@ -0,0 +1,314 @@
+            ? StepOption.kOverAsyncSuspension
+            : StepOption.kOver);
+  }
+
+  Future<Success> stepIn() {
+    return _service.resume(isolateRef.id, step: StepOption.kInto);
+  }
+
+  Future<Success> stepOut() {
+    return _service.resume(isolateRef.id, step: StepOption.kOut);
+  }
+
+  Future<void> clearBreakpoints() async {
+    final List<Breakpoint> breakpoints = _breakpoints.value.toList();
+    await Future.forEach(breakpoints, (Breakpoint breakpoint) {
+      return removeBreakpoint(breakpoint);
+    });
+  }
+
+  Future<void> addBreakpoint(String scriptId, int line) {
+    return _service.addBreakpoint(isolateRef.id, scriptId, line);
+  }
+
+  Future<void> addBreakpointByPathFragment(String path, int line) async {
+    final ScriptRef ref =
+        scripts.firstWhere((ref) => ref.uri.endsWith(path), orElse: () => null);
+    if (ref != null) {
+      return _service.addBreakpoint(isolateRef.id, ref.id, line);
+    }
+  }
+
+  Future<void> removeBreakpoint(Breakpoint breakpoint) {
+    return _service.removeBreakpoint(isolateRef.id, breakpoint.id);
+  }
+
+  Future<void> setExceptionPauseMode(String mode) {
+    return _service.setExceptionPauseMode(isolateRef.id, mode);
+  }
+
+  Future<Stack> getStack() async {
+    final stack = await _service.getStack(isolateRef.id);
+    if (stack is Sentinel) return null;
+    return stack;
+  }
+
+  InstanceRef get reportedException => _reportedException;
+
+  void _handleIsolateEvent(Event event) {
+    if (event.isolate.id != isolateRef.id) {
+      return;
+    }
+
+    _hasFrames.value = event.topFrame != null;
+    lastEvent = event;
+
+    switch (event.kind) {
+      case EventKind.kResume:
+        _isPaused.value = false;
+        _reportedException = null;
+        break;
+      case EventKind.kPauseStart:
+      case EventKind.kPauseExit:
+      case EventKind.kPauseBreakpoint:
+      case EventKind.kPauseInterrupted:
+      case EventKind.kPauseException:
+      case EventKind.kPausePostRequest:
+        _reportedException = event.exception;
+        _isPaused.value = true;
+        break;
+      case EventKind.kBreakpointAdded:
+        _breakpoints.value = [..._breakpoints.value, event.breakpoint];
+        break;
+      case EventKind.kBreakpointResolved:
+        _breakpoints.value = [
+          for (var b in _breakpoints.value) if (b != event.breakpoint) b,
+          event.breakpoint
+        ];
+        break;
+      case EventKind.kBreakpointRemoved:
+        _breakpoints.value = [
+          for (var b in _breakpoints.value) if (b != event.breakpoint) b
+        ];
+        break;
+    }
+  }
+
+  void _clearCaches() {
+    _scriptCache.clear();
+    lastEvent = null;
+    _reportedException = null;
+  }
+
+  void dispose() {
+    _debugSubscription?.cancel();
+  }
+
+  /// Get the populated [Instance] object, given an [InstanceRef].
+  ///
+  /// The return value can be one of [Instance] or [Sentinel].
+  Future<dynamic> getInstance(InstanceRef instanceRef) {
```

Comment:
I think we should be returning type `Obj` here, not dynamic

Reply:
what values other than Instance are returned?